lib/bolt/pal/yaml_plan.rb

```diff
@@ -67,82 +37,39 @@ def initialize(name, plan)
-                                  " #{invalid_keys.to_a.inspect}",
-                                  "bolt/invalid-plan")
-          end
-          type = Puppet::Pops::Types::TypeParser.singleton.parse(definition['type']) if definition.key?('type')
-          Parameter.new(param, definition['default'], type)
-        end.freeze
+        unless plan['steps'].is_a?(Array)
+          raise Bolt::Error.new("Plan must specify an array of steps", "bolt/invalid-plan")
+        end
 
-        @steps = plan['steps']&.map do |step|
+        used_names = Set.new(@parameters.map(&:name))
+
+        @steps = plan['steps'].each_with_index.map do |step, index|
+          duplicate_check(used_names, step['name'], index + 1)
+          used_names << step['name'] if step['name']
           # Step keys also aren't allowed to be code and neither is the value of "name"
           stringified_step = Bolt::Util.walk_keys(step) { |key| stringify(key) }
           stringified_step['name'] = stringify(stringified_step['name']) if stringified_step.key?('name')
-          stringified_step
+          Step.new(stringified_step, index + 1)
         end.freeze
-
         @return = plan['return']
-
-        validate
       end
 
-      VAR_NAME_PATTERN = /\A[a-z_][a-z0-9_]*\z/.freeze
-
-      def validate
-        unless @steps.is_a?(Array)
-          raise Bolt::Error.new("Plan must specify an array of steps", "bolt/invalid-plan")
-        end
-
-        used_names = Set.new
-        step_number = 1
-
-        # Parameters come in a hash, so they must be unique
-        @parameters.each do |param|
-          unless param.name.is_a?(String) && param.name.match?(VAR_NAME_PATTERN)
-            raise Bolt::Error.new("Invalid parameter name #{param.name.inspect}", "bolt/invalid-plan")
-          end
-
-          used_names << param.name
-        end
-
-        @steps.each do |step|
-          validate_step_keys(step, step_number)
-
-          begin
-            step.each { |k, v| validate_puppet_code(k, v) }
-          rescue Bolt::Error => e
-            raise Bolt::Error.new(step_err_msg(step_number, step['name'], e.msg), 'bolt/invalid-plan')
-          end
-
-          if step.key?('name')
-            unless step['name'].is_a?(String) && step['name'].match?(VAR_NAME_PATTERN)
-              error_message = "Invalid step name: #{step['name'].inspect}"
-              raise Bolt::Error.new(step_err_msg(step_number, step['name'], error_message), "bolt/invalid-plan")
-            end
-
-            if used_names.include?(step['name'])
-              error_message = "Duplicate step name or parameter detected: #{step['name'].inspect}"
-              raise Bolt::Error.new(step_err_msg(step_number, step['name'], error_message), "bolt/invalid-plan")
-            end
-
-            used_names << step['name']
-          end
-          step_number += 1
+      def duplicate_check(used_names, name, index)
```

Comment:
This does not seem to be detecting duplicate step names.

Reply:
Ugh, that might actuallly be a hard fix. Do you have the plan that's failing?

Reply:
Should be any plan with duplicate step names. Is it not failing for you?